sql-tuning-best-practice.md

```diff
@@ -0,0 +1,738 @@
+
+
+```SQL
+EXPLAIN SELECT COUNT(*) FROM trips WHERE start_date BETWEEN '2017-07-01 00:00:00' AND '2017-07-01 23:59:59';
+
++--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+
+| id                       | estRows     | task         | access object     | operator info                                                                                      |
++--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+
+| StreamAgg_20             | 1.00        | root         |                   | funcs:count(Column#13)->Column#11                                                                  |
+| └─TableReader_21         | 1.00        | root         |                   | data:StreamAgg_9                                                                                   |
+|   └─StreamAgg_9          | 1.00        | cop[tikv]    |                   | funcs:count(1)->Column#13                                                                          |
+|     └─Selection_19       | 250.00      | cop[tikv]    |                   | ge(trips.start_date, 2017-07-01 00:00:00.000000), le(trips.start_date, 2017-07-01 23:59:59.000000) |
+|       └─TableFullScan_18 | 10000.00    | cop[tikv]    | table:trips       | keep order:false, stats:pseudo                                                                     |
++--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+
+5 rows in set (0.00 sec)
+```
+
+Let's apply the "first child first – recursive descent" rule to the second plan. In the below example begin from the top to bottom, by looking at the `IndexRangeScan_47` (the first child of the tree). For the table `stars`, the optimizer ony need to select the column `name` and `id`, the two columns can be met by the index `name(name)`. So for the table `star`, the root reader is `IndexReader_48`, rather than a `TableReader`.
+The join method between `stars` and `planets` is a hash join, which is marked as `HashJoin_44`. The data access method on `planets` is a `TableFullScan_45`. After the join, the `TopN_26` and `TOPN_19` is to implemented the two order by and limit corespondingly. The final operator `Projection_16` is to implemented the column projection for `t5.name`.
+
+
+```SQL
+tidb> EXPLAIN SELECT t5.name FROM
+    -> (SELECT p.name, p.gravity, p.distance_from_sun FROM universe.planets p JOIN universe.stars s
+    ->  ON s.id = p.sun_id AND s.name = 'Sun'
+    ->  ORDER BY p.distance_from_sun ASC LIMIT 5) t5
+    -> ORDER BY t5.gravity DESC LIMIT 3;
++-----------------------------------+----------+-----------+---------------------------+
```

Comment:
````suggestion
```sql
EXPLAIN SELECT t5.name FROM
(SELECT p.name, p.gravity, p.distance_from_sun FROM universe.planets p JOIN universe.stars s
ON s.id = p.sun_id AND s.name = 'Sun'
ORDER BY p.distance_from_sun ASC LIMIT 5) t5
ORDER BY t5.gravity DESC LIMIT 3;
```

```
+-----------------------------------+----------+-----------+---------------------------+
````